### PR TITLE
release: build and push Docker image

### PR DESCRIPTION
**Pull request by liubin**
When new release is created, build and push Docker image to GitHub Packages.

Fixe: #237 

Signed-off-by: bin liu <liubin0329@gmail.com>

### CODE REVIEW

File: misc/snapshotter/Dockerfile [hunk 1]
```diff
@@ -3,20 +3,33 @@ FROM ubuntu:20.04 AS sourcer
...
     apt install --no-install-recommends -y ca-certificates && \
     rm -rf /var/cache/apt/* /var/lib/apt/lists/*
+
+VOLUME /var/lib/containerd-nydus /run/containerd-nydus
+
+COPY --from=sourcer /.nydus_version /.nydus_version
+
+## FIXME: how to pass nydus snapshotter version?
+## use: build-args?
+ENV NYDUS_SNAPSHOTTER_VERSION=vx.y.z
```
[the remark this answers is not in the document]
Comment (liubin) [reply misc/snapshotter/Dockerfile:24]: Indeed if we use the git tag/branch as the image tag, this version can be removed.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -70,3 +72,34 @@ jobs:
...
             ${{ env.tarball_shasum }}
+
+  publish-image:
+    runs-on: ubuntu-latest
+    needs: [build]
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Log in to the container registry
+        uses: docker/login-action@f054a8b539a109f9f41c372932f1ae047eff08c9
```
[the remark this answers is not in the document]
Comment (changweige) [reply .github/workflows/release.yml:83]: Seems we can move to `v2` right now. 
https://github.com/docker/login-action#docker-hub
```
jobs:
  login:
    runs-on: ubuntu-latest
    steps:
      -
        name: Login to GitHub Container Registry
        uses: docker/login-action@v2
        with:
          registry: ghcr.io
          username: ${{ github.actor }}
          password: ${{ secrets.GITHUB_TOKEN }}
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -70,3 +72,34 @@ jobs:
...
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: download artifacts
+        uses: actions/download-artifact@v2
+        with:
+          name: nydus-snapshotter_artifacts
+          path: misc/snapshotter
+      - name: Extract metadata (tags, labels) for Docker
+        id: meta
+        uses: docker/metadata-action@98669ae865ea3cffbcbaa878cf57c20bbf1c6c38
```
[the remark this answers is not in the document]
Comment (changweige) [reply .github/workflows/release.yml:95]:
```
    runs-on: ubuntu-latest
    steps:
      -
        name: Checkout
        uses: actions/checkout@v3
      -
        name: Docker meta
        id: meta
        uses: docker/metadata-action@v4
        with:
          images: name/app

```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -70,3 +72,34 @@ jobs:
...
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+      - name: build and push nydus-snapshotter image
+        uses: docker/build-push-action@ad44023a93711e3deb337508980b4b5e9bcdc5dc
```
[the remark this answers is not in the document]
Comment (changweige) [reply .github/workflows/release.yml:99]:
```
        with:
          username: ${{ secrets.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}
      -
        name: Build and push
        uses: docker/build-push-action@v3
        with:
          context: .
          push: ${{ github.event_name != 'pull_request' }}
          tags: ${{ steps.meta.outputs.tags }}
          labels: ${{ steps.meta.outputs.labels }}

```

Please refer to https://github.com/docker/metadata-action#usage

File: misc/snapshotter/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (changweige) [reply misc/snapshotter/Dockerfile:24]: Then let's remove it? It is not referenced anywhere

File: misc/snapshotter/entrypoint.sh [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
-CONTAINERD_ROOT="${CONTAINERD_ROOT:-/var/lib/containerd/}"
+NYDUS_LIB="${NYDUS_LIB:-/var/lib/containerd-nydus}"
+NYDUS_RUN="${NYDUS_RUN:-/run/containerd-nydus}"
+ENABLE_NYDUS_OVERLAY="${ENABLE_NYDUS_OVERLAY:-true}"
+ENABLE_METRICS="${ENABLE_METRICS:-false}"
+LEVEL="${LEVEL:-info}"
 
 set -eu
 BACKEND_TYPE="${BACKEND_TYPE:-config}"
 NYDUSD_DAEMON_MODE="${NYDUSD_DAEMON_MODE:-shared}"
```
Comment (changweige) [inline misc/snapshotter/entrypoint.sh:11]: Not relating to this PR, share we change the daemon mode from `shared` to `multiple`?

File: misc/snapshotter/Dockerfile [hunk 6]
```diff
@@ -3,20 +3,33 @@ FROM ubuntu:20.04 AS sourcer
...
+
+RUN mkdir -p /usr/local/bin/ /etc/nydus/ /var/lib/containerd-nydus/cache /tmp/blobs/
+COPY --from=sourcer /nydus* /usr/local/bin/
+# FIXME: need an universal method to get containerd-nydus-grpc binary
+COPY containerd-nydus-grpc /usr/local/bin/
```
Comment (changweige) [inline misc/snapshotter/Dockerfile:29]: Does `containerd-nydus-grpc` exist in `misc/snapshotter` ?

File: .github/workflows/release.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (liubin) [reply .github/workflows/release.yml:83]: I'll fix this later, now I'm testing it.

File: misc/snapshotter/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (liubin) [reply misc/snapshotter/Dockerfile:24]: Yes, I'll remove it.

File: misc/snapshotter/entrypoint.sh (hunk 5 above)
Comment (liubin) [reply misc/snapshotter/entrypoint.sh:11]: @imeoer your option?

File: misc/snapshotter/Dockerfile (hunk 6 above)
Comment (liubin) [reply misc/snapshotter/Dockerfile:29]: No, that is why I added a FIXME.
The `containerd-nydus-grpc` binary is download by [the `download artifacts` step](https://github.com/containerd/nydus-snapshotter/pull/238/files#diff-87db21a973eed4fef5f32b267aa60fcee5cbdf03c67fafdc2a9b553bb0b15f34R88-R92).

Indeed this line is unchanged. IMO we can improve it by add an `image` target in the Makefile to help users to not copy the binary manually.

Comment (changweige) [reply misc/snapshotter/Dockerfile:29]: SGTM

**Pull request merged**